compat/sched.h

```diff
@@ -1,31 +1,37 @@
 /*
- * Copyright (C) 1996-2020 The Squid Software Foundation and contributors
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
  *
  * Squid software is distributed under GPLv2+ license and includes
  * contributions from numerous individuals and organizations.
  * Please see the COPYING and CONTRIBUTORS files for details.
  */
 
-#ifndef SQUID_COMPAT_CPU_H
-#define SQUID_COMPAT_CPU_H
+#ifndef SQUID_COMPAT_SCHED_AFFINITY_H
+#define SQUID_COMPAT_SCHED_AFFINITY_H
 
-#if HAVE_ERRNO_H
-#include <errno.h> /* for ENOTSUP */
+#if !HAVE_SCHED_SETAFFINITY
+
+#if HAVE_STDBOOL_H
+#include <stdbool.h>
 #endif
-#if HAVE_SCHED_H
-#include <sched.h>
+#if HAVE_SYS_PARAM_H
+#include <sys/param.h>
 #endif
-
-#if !HAVE_CPU_AFFINITY
-/* failing replacements to minimize the number of if-HAVE_CPU_AFFINITYs */
-#if !HAVE_CPU_SET_T
+#if HAVE_SYS_CPUSET_H
+#include <sys/cpuset.h>
+typedef cpuset_t cpu_set_t;
+// CPU_AND needs a little reimplementation due to signature difference and the key difference is it s implemented as a do while loop
+#undef CPU_AND
+#define CPU_AND(destset, srcset1, srcset2) CpuAnd((destset), (srcset1), (srcset2))
```

Comment:
This PR provides replacements for the Linux-standard sched_getaffinity family of functions using a _partially overlapping_ but different FreeBSD-standard family of functions. I am pretty sure that the overlaps cannot be _safely_ handled by redefining standard macros. Sooner or later, some FreeBSD header will mention the macro will redefined, screwing Squid.

The correct solution here is, IMO, elimination of all CPU_AND() calls from Squid code outside of this compatibility shim. That code should use CpuAnd() and this shim should provide that function in all cases, including Linux. The Linux version will implement it using CPU_AND (where available), of course.